src/TranslatableTrait.php

```diff
@@ -0,0 +1,74 @@
+<?php
+
+namespace atk4\core;
+
+/**
+ * If a class use this trait, string can be translated calling method translate.
+ */
+trait TranslatableTrait
+{
+    /**
+     * Check this property to see if trait is present in the object.
+     *
+     * @var bool
+     */
+    public $_translatableTrait = true;
+
+    /** @var TranslatorInterface */
+    protected $translator;
+
+    /**
+     * Set Translator.
+     *
+     * @TODO should be setted internally or leave it public and leave it definable via Defaults?
+     *
+     * @param TranslatorInterface $translator
+     */
+    public function setTranslator(TranslatorInterface $translator)
+    {
+        $this->translator = $translator;
+    }
+
+    /**
+     * Translate a string.
+     *
+     * @param string $string
+     * @param mixed  ...$args
+     *
+     * @throws Exception
+     *
+     * @return string
+     */
+    public function __(string $string, ...$args)
+    {
+        $count_args = count($args);
+        // translation : simple case
+        if (0 === $count_args) {
+            return $this->translator->translate($string);
+        }
+
+        // translation : plural case
+        if (1 === count($args) && is_numeric($args[0])) {
+            return $this->translator->translate_plural($string, $args[0]);
+        }
+
+        $translated_args = [];
+        $translated_args[] = $this->translator->translate($string);
+        foreach ($args as $sub_args) {
+            if (!is_array($sub_args)) {
+                $sub_args = [$sub_args];
+            }
+
+            $sub_string = $sub_args[0];
+            $sub_args = $sub_args[1] ?? false;
+
+            if (!$sub_args) {
+                $translated_args[] = $this->__($sub_string);
+            } else {
+                $translated_args[] = $this->__($sub_string, $sub_args);
+            }
+        }
+
+        return call_user_func_array('sprintf', $translated_args);
```

Comment:
1. use TransatableTrait
2. $this->__('we add %s into %s', ['milk', 'bowl']);

^^ the only thing trait should do is check for API and fallback to sprintf..

Reply:
do it but i add two helper method that make the usage of the trait much short and declarative with array. look in this test for helper usage :
https://github.com/abbadon1334/core/blob/dc07b17f949fdea4f42c7b2b56351b5dd15736e1/tests/TranslatableTraitAppTest.php#L271

Reply:
ok, can we use `$this->__('we add {thing} into {container}', ['thing'=>'milk', 'container'=>'bowl'])` ?

That should be compatible with http://userguide.icu-project.org/formatparse/messages.